### ISSUE TITLE
Frontend protocol: why cursor updates are bundled with text updates?

### ISSUE THREAD

**Issue by c4eater** [xi-editor/xi-editor#1228]
Hi,

I think I probably found a behavioral issue in the frontend protocol.

The question is, why does the protocol bundle the cursor updates with text updates?
For instance, here are the notifications sent by xi-core when I move the cursor down from line "mmm" to line "nnn". (The client is, quite expected, using a line cache that is storing a contiguous vector of recently rendered lines printed to the client window)

 1. copy n=12 (instructs to copy 12 lines from the old cache before "mmm". 12 is just some irrelevant # of lines before the cursor)
   
 2. ins n=2 (instructs to insert 2 new lines to the cache)
   
  * line: mmm
  * line: nnn cursor=[0]

 3. skip n=2 (instructs to skip 2 lines from the old cache, which I guess are "mmm"+cursor and "nnn")

 4. copy n = 33 (copy remaining lines from the old cache)

Effectively, this bunch of operations notifies the client of a cursor move. The text itself remains intact.
But there is no direct way for the client to know that the text hasn't been changed, unless the client starts explicitly comparing the skipped old lines with the incoming lines (which is ineffective).

Btw the frontend spec has an "update" operation which can move the cursor, but says it's currently unused. Why?

I'd add that my client is a console client, thus the cursor redrawing is performed as an individual operation after redrawing the text itself, and is done by simple termion::cursor::Goto. Thus, it's quite natural to expect cursor motion notifications as individual updates. Here is what I would expect from xi-core:

 1. copy n=45 (instructs to copy 45 lines from the old cache - this is the whole screen)
   
 2. update (imaginary operator instructing to reposition the cursor)
   
Thanks in advance. If that is needed I'll write a simple test that reproduces the aforementioned workflow.

**Comment by cmyr** [comment]
The purpose of the update protocol & the line cache is not to provide information or metadata about the document; it is only to give the frontend the information required to draw the currently visible region.

For our purposes, it is useful to think of the frontend as a 'dumb terminal'. All it does is draw.

The "update" operation was intended as a little efficiency win to reduce how much data we send when just styles or cursors changed, but it wasn't implemented because the initial implementation had the performance characteristics we wanted, and so it wasn't a priority. In any case, this was not intended to provide information to the frontend, it was just to reduce the amount of data sent during updates.

Basically: the frontend isn't supposed to know anything. It's just supposed to send user events to core, and get back the current document state, so that it can draw. If you wanted to actually know what the user was doing, that was the domain for plugins; and there was discussion about unifying these things into a single protocol, but that work has not begun. See #845 for some discussion of this.

**Comment by c4eater** [comment]
Well, I'm not saying the client should know anything about the document state/metadata.

The flag "cache lines have been modified: y/n" is a characteristic of the cache lines. There is no encapsulation leakage here.

Both updates "copy 47, move cursor" and "copy 12, replace 2, copy 33" would provide the same cache and cursor state in the end, but in the former case, the client would know that the cache content was not updated. This can save a redraw or multiple redraws. Consider a xi-term issue I'm referencing a few comments earlier: the cache (which is rather simplistic at the moment; I'm talking of xrl::LineCache btw) only knows its current contents at the moment, which leads to frequent full redraws and screen flickering.

Yet, I'm going to close this issue shortly because:

1) it can be considered a client-side issue: a good cache struct on the frontend side should be capable of cache invalidation at a line-level granularity; that would make the modification of 2 lines a cheap operation anyway. I'm actually going to add this invalidation support (= "dirty" flag on cache lines) on xrl side.
2) from your reply, my understanding is that the cursor movements via "update" is a potential todo item, so no need to have a separate issue for that.

**Comment by cmyr** [comment]
Basically I think we're in agreement; it would be very reasonable for the protocol to special-case the situation where only cursors/styles have changed; this has been planned; it just never was hurting us enough to get implemented.

I do agree that it sounds like the frontend has some available improvements that aren't implemented; in xi-mac we redraw the smallest region required based on the update. In the case of a single cursor moving between lines, we only need to redraw those two lines.
